### PR TITLE
New utils/cube methods cube_squeeze, demote_dim_coord_to_aux_coord, promote aux_coord_to_dim_coord

### PR DESCRIPTION
**Pull request by kwilliams-mo**
Submitting for preliminary code review.
I wasn't really sure where these utilities might fit, so I put cube_squeeze as a util (since e.g. new_axis is a util) and demote_dim_coord_to_aux_coord, promote aux_coord_to_dim_coord as cube methods (since e.g. remove_coord, add_dim_coord etc are cube methods).

### CODE REVIEW

**Comment by rhattersley** [comment]
Thanks @kwilliams-mo - much appreciated. :smile:

I'm inclined to tweak the API a little (i.e. the routine names and where they live) but the functionality itself is in a really good state.

For the API:

1. How about having them all as functions instead of methods? We don't have a consistent view on functions vs. methods but maybe now is a good time to come up with one? (I'll raise this on the [iris-dev group](https://groups.google.com/forum/#!forum/scitools-iris-dev) to try to avoid polluting this PR with all that discussion.)
2. Are people happy with "demote" vs "promote"?

File: lib/iris/cube.py [hunk 1]
```diff
@@ -991,6 +991,157 @@ def replace_coord(self, new_coord):
...
+                   "Ensure all coordinates are of type basestring or "
+                   "iris.coords.Coord.")
+            msg = msg.format(type(name_or_coord))
+            raise TypeError(msg)
+
+        if aux_coord in self.dim_coords:
+            # nothing to do
+            return
+
+        if not aux_coord in self.aux_coords:
```
Comment (rhattersley) [inline lib/iris/cube.py]: The recommended form for this is `if aux_coord not in self.aux_coords:`. This is what's causing the PEP8 checker to fail.

File: lib/iris/cube.py [hunk 2]
```diff
@@ -991,6 +991,157 @@ def replace_coord(self, new_coord):
...
+            msg = msg.format(type(name_or_coord))
+            raise TypeError(msg)
+
+        if not dim_coord in self.dim_coords:
+            # nothing to do
+            return
+
+        coord_dim = self.coord_dims(dim_coord)
+
+        aux_coord = iris.coords.AuxCoord.from_coord(dim_coord)
```
Comment (rhattersley) [inline lib/iris/cube.py]: (Confusingly!) You don't need to convert a DimCoord to an AuxCoord in order to add it to the aux_coords. Keeping it as a DimCoord would preserve the knowledge that it is monotonic.

In other words, I'd just do:
```python
self.remove_coord(dim_coord)
self.add_aux_coord(dim_coord, coord_dim)
```

File: lib/iris/cube.py [hunk 3]
```diff
@@ -991,6 +991,157 @@ def replace_coord(self, new_coord):
...
+            dim_coord = name_or_coord
+        else:
+            # Don't know how to handle this type
+            msg = ("Don't know how to handle coordinate of type {}. "
+                   "Ensure all coordinates are of type basestring or "
+                   "iris.coords.Coord.")
+            msg = msg.format(type(name_or_coord))
+            raise TypeError(msg)
+
+        if not dim_coord in self.dim_coords:
```
Comment (rhattersley) [inline lib/iris/cube.py]: As elsewhere, `not <thing> in <other-thing>` is preferred as `<thing> not in <other-thing>`.

File: lib/iris/util.py [hunk 4]
```diff
@@ -1211,6 +1211,43 @@ def add_coord(coord):
     return new_cube
 
 
+def cube_squeeze(cube):
```
Comment (ajdawson) [inline lib/iris/util.py]: Could just call this `squeeze`, its presence in iris indicates it operates on cubes (as does the docstring).

File: lib/iris/util.py [hunk 5]
```diff
@@ -1211,6 +1211,43 @@ def add_coord(coord):
...
+
+        >>> cube.shape
+        (1, 360, 360)
+        >>> ncube = iris.util.cube_squeeze(cube)
+        >>> ncube.shape
+        (360, 360)
+
+    .. warning::
+
+        Calling this method will trigger any deferred loading, causing the
```
Comment (ajdawson) [inline lib/iris/util.py]: Is that true? I thought cube indexing preserved the deferred state of the data within.

File: lib/iris/cube.py [hunk 6]
```diff
@@ -991,6 +991,157 @@ def replace_coord(self, new_coord):
...
+
+        coord_dim = self.coord_dims(aux_coord)
+
+        if len(coord_dim) != 1:
+            msg = ("Attempting to promote an AuxCoord ({}) "
+                   "which is associated with {} dimensions")
+            msg = msg.format(aux_coord.name(), len(coord_dim))
+            raise ValueError(msg)
+
+        dim_coord = iris.coords.DimCoord.from_coord(aux_coord)
```
Comment (ajdawson) [inline lib/iris/cube.py]: What happens if this fails, and is the message in the exception clear for someone who just called *this* function?

**Comment by kwilliams-mo** [comment]
Thanks for the feedback @rhattersley and @ajdawson! Hopefully the latest commit should have all your suggested edits.

**Comment by rhattersley** [comment]
Thanks for the nudge @kwilliams-mo - it can be hard to spot new code being pushed to a PR without a corresponding comment. I'll add this to our v1.8 milestone.

**Comment by bjlittle** [comment]
:+1: Great work @kwilliams-mo! Thanks for this PR :smile:

**Pull request merged**